mchf-eclipse/drivers/ui/ui_driver.c

```diff
@@ -6940,6 +6960,82 @@ void UiDriver_TaskHandler_HighPrioTasks()
 
 }
 
+static void UiDriver_HandleUSB_Keyboard()
+{
+#ifdef USE_USBKEYBOARD
+    HID_KEYBD_Info_TypeDef *k_pinfo;
+    static uint8_t keys_buffer[sizeof(k_pinfo->keys)] = { 0 };
+
```

Comment:
> Good job! But one serious issue: You are modifying STM_HAL files. This is not recommended, as it impairs our ability to update the STM_HAL easily. I know the change is small but can we do it in a different way?

Did you find some way to not modify STM stuff? I didn't, maybe put USB handling into uhsdr_usb_kb.c, copy stuff from stm where? anyway looks like handling USB will grow, I would like to add some more features, change CW speed, change band, maybe mode, recording F-bottuns on the long press, maybe users suggest something more.... @db4ple , What do you think?

Reply:
Yes, I think I do. I thought I already posted it. We have two options:
Reimplemented the usbh_keybd.c code (one for all platforms, we did that for our composite usb device already).

Or: We simply copy the character of interest to the first location in the array and call the unmodified function. Which is simpler and would do the job for now. 

General remark: Keyboards are a pain because the USB spec did not specific a keyboard layout identification. It feels so much eigthies to have to select the proper layout manually...

Reply:
> Yes, I think I do. I thought I already posted it. We have two options:
> Reimplemented the usbh_keybd.c code (one for all platforms, we did that for our composite usb device already).
> 
> Or: We simply copy the character of interest to the first location in the array and call the unmodified function. Which is simpler and would do the job for now.
> 
> General remark: Keyboards are a pain because the USB spec did not specific a keyboard layout identification. It feels so much eigthies to have to select the proper layout manually...

The second one is a mess. I will take a look at the first one....